### PR TITLE
Rewrite patch

### PR DESCRIPTION
**Pull request by okwolf**
Lots of inspiration from [Ultradom](https://github.com/jorgebucaran/ultradom), but with a very different VDOM.

Added:
* Components were already lazy, but now they have access to `state` and `fx` like the top level view, making them full subviews
* Multiple root nodes allowed

Intentionally not included:

* Hydration
* CSS variables
* Keyed updates

Future TODO:

* Lifecycle events other than `oncreate`
* Fragments (already supported at the root level)

### CODE REVIEW

File: src/patch.js [hunk 1]
```diff
@@ -21,48 +23,91 @@ export function patch(node, container, runFx) {
       } else if (
         value != null &&
         value !== false &&
-        lifecycleEvents.indexOf(name) === -1
+        lifecycleEventNames.indexOf(name) === -1
```
Comment (jorgebucaran) [inline src/patch.js:25]: No idea what's going on here! 😄

File: src/patch.js [hunk 2]
```diff
@@ -21,48 +23,91 @@ export function patch(node, container, runFx) {
...
-            nextElement.appendChild(childElement);
+  function createElement(node, propsNode) {
+    var element =
+      typeof node === "string" || typeof node === "number"
+        ? document.createTextNode(node)
+        : document.createElement(node[0]);
+    element.vnode = node;
+    stack.push(function() {
+      // oncreate
+      runFx((propsNode[1] || {})[lifecycleEventNames[0]], element);
```
Comment (jorgebucaran) [inline src/patch.js]: @okwolf Interesting. Are lifecycle events managed effects? How would this work?

File: src/patch.js (hunk 1 above)
Comment (okwolf) [reply src/patch.js:25]: Avoiding `<div oncreate="function () { ... } ">`.

File: src/patch.js (hunk 2 above)
Comment (okwolf) [reply src/patch.js]: Yes. `@hyperapp/fx` supports this already using `event` fx. This implementation is based on the same premise, but built-in.

File: src/patch.js (hunk 1 above)
Comment (jorgebucaran) [reply src/patch.js:25]: Hmm.

File: src/patch.js (hunk 2 above)
Comment (jorgebucaran) [reply src/patch.js]: https://github.com/hyperapp/fx#event

Comment (jorgebucaran) [reply src/patch.js]: @okwolf I need to review that documentation to see why is this useful.

Comment (okwolf) [reply src/patch.js]: The only limitation of `event` is that it only handles unary event handler functions, which for Hyperapp lifecycle events means only access to the `element`.

**Pull request merged**